File: 155-min-stack/min-stack.js

var MinStack = function () {
    this.min = [];
    this.stack = [];
};

/** 
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function (val) {
    if (!this.stack.length) this.min.push({ value: val, index: 0 });
    else if (val <= this.min[this.min.length - 1].value) this.min.push({ value: val, index: this.stack.length });
    this.stack.push(val)
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function () {
    if (this.stack.length - 1 === this.min[this.min.length - 1].index) this.min.pop();
    return this.stack.pop();
};

/**
 * @return {number}
 */
MinStack.prototype.top = function () {
    return this.stack[this.stack.length - 1];
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function () {
    return this.min[this.min.length-1].value;
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: 155-min-stack/min-stack.cs
public class MinStack {

    private Stack<int> stack;
    private Stack<int> minValue;

    public MinStack() {
        this.stack = new Stack<int>();    
        this.minValue = new Stack<int>();
    }
    
    public void Push(int val) {
        this.stack.Push(val);
        if(this.stack.Count == 1) minValue.Push(val);
        else if(this.minValue.Peek() >= val) this.minValue.Push(val);
    }
    
    public void Pop() {
        if(this.stack.Count == 0) return;
        if(this.minValue.Peek() == this.stack.Pop()) this.minValue.Pop();
    }
    
    public int Top() {
        if(this.stack.Count != 0) return this.stack.Peek();
        else throw new InvalidOperationException();
    }
    
    public int GetMin() {
        if(this.minValue.Count != 0) return this.minValue.Peek();
        else throw new InvalidOperationException();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.Push(val);
 * obj.Pop();
 * int param_3 = obj.Top();
 * int param_4 = obj.GetMin();
 */